### PR TITLE
Feat: Numbers 1.0.2 & Versioned V1 usage

### PR DESCRIPTION
**Pull request by JPPortier**

### CODE REVIEW

File: client/src/main/com/sinch/sdk/domains/numbers/adapters/converters/ScheduledVoiceProvisioningDtoConverter.java [hunk 1]
```diff
@@ -3,34 +3,45 @@
...
-  public static ScheduledVoiceProvisioning convert(ScheduledVoiceProvisioningDto dto) {
+  public static ScheduledVoiceProvisioning convert(
+      com.sinch.sdk.domains.numbers.models.v1.ScheduledVoiceProvisioning _dto) {
 
-    if (null == dto) {
+    if (null == _dto) {
       return null;
     }
+
+    if (!(_dto instanceof ScheduledVoiceProvisioningRTC)) {
```
Comment (asein-sinch) [inline client/src/main/com/sinch/sdk/domains/numbers/adapters/converters/ScheduledVoiceProvisioningDtoConverter.java:17]: Is it for backward compatibility as other types of voice configuration are not handled by the previous specification version?

File: client/src/main/com/sinch/sdk/domains/numbers/api/v1/AvailableNumberService.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+   * <p>Note: You cannot add both objects if you only need to configure one object. For example, if
+   * you only need to configure smsConfiguration for SMS messaging, do not add the
+   * voiceConfiguration object or it will result in an error.
+   *
+   * @param phoneNumber Number to be activated
+   * @param parameters Activation parameters
+   * @return Activated number
+   * @since 1.2
+   */
+  ActiveNumber rent(String phoneNumber, AvailableNumberRentRequest parameters) throws ApiException;
```
Comment (asein-sinch) [inline client/src/main/com/sinch/sdk/domains/numbers/api/v1/AvailableNumberService.java]: Wild question: I've always been puzzled by the overlap of `available` and `active`.
What do you think of the following in order to improve the user experience: merge both subdomains and change the functions naming:
 - numbers().available().list() -> number().searchForAvailableNumbers()
 - numbers().available().checkAvailability() -> numbers().checkAvailability()
 - numbers().available().rent() -> numbers().rent()
 - numbers().available().rentAny() -> numbers().rentAny()
 - numbers().active().list() -> numbers().list()
 - numbers().active().get() -> numbers().get()
 - numbers().active().update() -> numbers().update()
 - numbers().active().release() -> numbers().release()
 
With the approach, the active numbers (the ones the user will see in the dashboard) become first class citizens in the API and the pool of available numbers is only present in the methods with long and descriptive names.
This is a wild question, subject to reflexion and discussion, I don't expect a definitive answer here 🙂

File: client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  public NumbersService(
+      UnifiedCredentials credentials, NumbersContext context, HttpClient httpClient) {
+
+    Objects.requireNonNull(
+        credentials, "Numbers service require unified credentials to be defined");
+    Objects.requireNonNull(context, "Numbers service require context must be defined");
+    StringUtil.requireNonEmpty(
+        credentials.getKeyId(), "Numbers service require  'keyId' must be defined");
```
Comment (asein-sinch) [inline client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java]: Review syntax and formatting (double space).
Suggestion:
`Numbers service requires 'keyId' to be defined`
or
`Numbers service's required property 'keyId' must be defined`

File: client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  public AvailableRegionService regions() {
+    if (null == this.regions) {
+      this.regions = new AvailableRegionService(uriUUID, context, httpClient, authManagers);
+    }
+    return this.regions;
+  }
+
+  public com.sinch.sdk.domains.numbers.api.v1.ActiveNumberService active() {
```
Comment (asein-sinch) [inline client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java]: Why do you need the fully qualified name to define the return type?

File: client/src/main/com/sinch/sdk/domains/numbers/models/v1/active/request/ActiveNumberListRequest.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @see <a
+ *     href="https://developers.sinch.com/docs/numbers/api-reference/numbers/tag/Active-Number">https://developers.sinch.com/docs/numbers/api-reference/numbers/tag/Active-Number/</a>
+ * @since 1.2
+ */
+public interface ActiveNumberListRequest {
+
+  /**
+   * @return Region code to filter by. ISO 3166-1 alpha-2 country code of the phone number. Example:
+   *     &#x60;US&#x60;, &#x60;GB&#x60; or &#x60;SE&#x60;. (required)
```
Comment (asein-sinch) [inline client/src/main/com/sinch/sdk/domains/numbers/models/v1/active/request/ActiveNumberListRequest.java]: Check characters encoding for javadoc

File: client/src/test/java/com/sinch/sdk/domains/numbers/adapters/ActiveNumberServiceTest.java [hunk 6]
```diff
@@ -88,42 +117,71 @@ void list() throws ApiException {
...
+                        .setProjectId("project id")
+                        .setRegionCode("GB")
+                        .setType(NumberType.MOBILE)
+                        .setCapability(Arrays.asList(Capability.SMS, Capability.VOICE))
+                        .setDisplayName("")
+                        .setMoney(new Money("EUR", 0.80))
+                        .setPaymentIntervalMonths(1)
+                        .setNextChargeDate(Instant.parse("2023-09-22T15:49:58.813424Z"))
+                        .setExpireAt(Instant.parse("2023-10-06T15:49:58.813381Z"))
+                        .setSmsConfiguration(new SMSConfiguration("service plan id", "", null))
```
Comment (asein-sinch) [inline client/src/test/java/com/sinch/sdk/domains/numbers/adapters/ActiveNumberServiceTest.java:133]: Shouldn't we use the builder instead of the constructor?

File: client/src/test/java/com/sinch/sdk/domains/numbers/adapters/AvailableNumberServiceTest.java [hunk 7]
```diff
@@ -80,37 +59,48 @@ void list() throws ApiException {
...
-                .setSupportingDocumentationRequired(true)
-                .build());
+        new ArrayList<>(
+            Collections.singletonList(
+                AvailableNumber.builder()
+                    .setPhoneNumber("+46650553763")
+                    .setRegionCode("SE")
+                    .setType(NumberType.LOCAL)
+                    .setCapability(new ArrayList<>(Collections.singletonList(Capability.VOICE)))
+                    .setSetupPrice(new Money("DOLLAR", 0.57))
```
Comment (asein-sinch) [inline client/src/test/java/com/sinch/sdk/domains/numbers/adapters/AvailableNumberServiceTest.java:68]: Should be `USD`

File: openapi-contracts/src/test/java/com/sinch/sdk/domains/numbers/models/v1/ScheduledVoiceProvisioningDtoTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  @GivenTextResource("/domains/numbers/v1/active/ScheduledVoiceProvisioningDtoFAX.json")
+  String provisioningJsonFAX;
+
+  @GivenTextResource("/domains/numbers/v1/active/ScheduledVoiceProvisioningDtoRTC.json")
+  String provisioningJsonRTC;
+
+  public static ScheduledVoiceProvisioningEST provisioningEST =
+      ScheduledVoiceProvisioningEST.builder()
+          .setTrunkId("EEEEEEEEEE-DDDD-CCCC-BBBB-AAAAAAAA")
```
Comment (asein-sinch) [inline openapi-contracts/src/test/java/com/sinch/sdk/domains/numbers/models/v1/ScheduledVoiceProvisioningDtoTest.java:28]: Nit: a trunkId or a serviceId are generally in the form of ULID

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoAppIdNull.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{
+  "appId": null,
```
Comment (asein-sinch) [inline openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoAppIdNull.json:2]: Can this happen?

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoRTCSwitchToFAX.json [hunk 10]
```diff
@@ -0,0 +1,11 @@
+{
+  "appId": "AAAAAAAA-BBBB-CCCC-DDDD-EEEEEEEEEE",
+  "scheduledVoiceProvisioning": {
```
Comment (asein-sinch) [inline openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoRTCSwitchToFAX.json:3]: For more realistic deserialization test, you can also add empty values for `trunkId` and `appId`

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+  "displayName": "a display",
+  "smsConfiguration": {
+    "servicePlanId": "service plan id",
+    "campaignId": "campaign id"
+  },
+  "voiceConfiguration": {
```
Comment (asein-sinch) [inline openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json]: you should also add 
"scheduledVoiceProvisioning": null,
"lastUpdatedTime": null,

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
+  "displayName": "a display",
+  "smsConfiguration": {
```
Comment (asein-sinch) [inline openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json]: you should also add 
"scheduledProvisioning": null,

File: client/src/main/com/sinch/sdk/domains/numbers/adapters/converters/ScheduledVoiceProvisioningDtoConverter.java (hunk 1 above)
Comment (JPPortier) [reply client/src/main/com/sinch/sdk/domains/numbers/adapters/converters/ScheduledVoiceProvisioningDtoConverter.java:17]: Exact: Non-versioned service is handling RTC only

File: client/src/main/com/sinch/sdk/domains/numbers/api/v1/AvailableNumberService.java (hunk 2 above)
Comment (JPPortier) [reply client/src/main/com/sinch/sdk/domains/numbers/api/v1/AvailableNumberService.java]: I agree with question 👍 
And agree too with conclusion:
> This is a wild question, subject to reflexion and discussion

I could add subjects related to:
-  `checkAvailability` behavior when return could be 'false' (return value vs dedicated exception)
- all SDKs should be synchronized if going to this direction

File: client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java (hunk 3 above)
Comment (JPPortier) [reply client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java]: Fixed (will do a Verification's dedicated PR with same fix)

File: client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java (hunk 4 above)
Comment (JPPortier) [reply client/src/main/com/sinch/sdk/domains/numbers/api/v1/adapters/NumbersService.java]: There was no need.
Fixed

File: client/src/main/com/sinch/sdk/domains/numbers/models/v1/active/request/ActiveNumberListRequest.java (hunk 5 above)
Comment (JPPortier) [reply client/src/main/com/sinch/sdk/domains/numbers/models/v1/active/request/ActiveNumberListRequest.java]: Fixed

File: client/src/test/java/com/sinch/sdk/domains/numbers/adapters/ActiveNumberServiceTest.java (hunk 6 above)
Comment (JPPortier) [reply client/src/test/java/com/sinch/sdk/domains/numbers/adapters/ActiveNumberServiceTest.java:133]: This is related to the non-versioned class... not based onto builder pattern (my bad)
Due to migration to new versioned usage, we could agree, I think there is no need to implement this builder.

File: client/src/test/java/com/sinch/sdk/domains/numbers/adapters/AvailableNumberServiceTest.java (hunk 7 above)
Comment (JPPortier) [reply client/src/test/java/com/sinch/sdk/domains/numbers/adapters/AvailableNumberServiceTest.java:68]: Used for unit test and because of a free form text, it should not hurt 😃 
Used values could be enhanced in the future if we need to stick to more real use case values

File: openapi-contracts/src/test/java/com/sinch/sdk/domains/numbers/models/v1/ScheduledVoiceProvisioningDtoTest.java (hunk 8 above)
Comment (JPPortier) [reply openapi-contracts/src/test/java/com/sinch/sdk/domains/numbers/models/v1/ScheduledVoiceProvisioningDtoTest.java:28]: Same as for USD vs DOLLAR

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoAppIdNull.json (hunk 9 above)
Comment (JPPortier) [reply openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoAppIdNull.json:2]: Historical test.
Not sure now with newly RTC default value as empty; but should not hurt to keep it alive

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoRTCSwitchToFAX.json (hunk 10 above)
Comment (JPPortier) [reply openapi-contracts/src/test/resources/domains/numbers/v1/active/VoiceConfigurationDtoRTCSwitchToFAX.json:3]: Right.
Done

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json (hunk 11 above)
Comment (JPPortier) [reply openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json]: This payload is used in a request context, so these fields do not have sense here
Renamed the file to highlight a "request" payload
Response payload is covered by the "GET" test because update is returning an AcitveNumber payload like a GET

File: openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json (hunk 12 above)
Comment (JPPortier) [reply openapi-contracts/src/test/resources/domains/numbers/v1/active/active-numbers-update.json]: (same as above: it is a request payload)

**Review by asein-sinch — APPROVED** [review]
👍
